docs/porting.md

```diff
@@ -239,33 +239,76 @@ def do_calculation(arg):
     if arg in global_cache:
         return global_cache[arg]
 
-    cache_lock.acquire()
-    if arg not in global_cache:
-        global_cache[arg] = _do_expensive_calculation(arg)
-    cache_lock.release()
+    with cache_lock:
+        if arg not in global_cache:
+            global_cache[arg] = _do_expensive_calculation(arg)
     return global_cache[arg]
 ```
 
 Note that after acquiring the lock, we first check if the requested key
 has been filled by another thread, to prevent unnecessary calls to
 `_do_expensive_calculation` if another thread filled the cache while the thread
-currently holding the lock was blocked on acquiring the lock. Also note that
+currently holding the lock was blocked on acquiring the lock. Also note that we
+avoid using
 [`Lock.acquire`](https://docs.python.org/3/library/threading.html#threading.Lock.acquire)
-*must* be followed by a call to
-[`Lock.release`](https://docs.python.org/3/library/threading.html#threading.Lock.release),
-calling `Lock.acquire()` recursively on the same lock leads to deadlocks. Also,
+and [`Lock.release`](https://docs.python.org/3/library/threading.html#threading.Lock.release)
+and instead we use the lock as a context manager. The difference is subtle: the context
+manager calls `Lock.release` in a `try...finally` clause, so if
```

Comment:
```suggestion
manager calls `Lock.release` in a `try` ... `finally` clause, so if
```

The CPython docs use `` :keyword:`try` ... :keyword:`finally` ``, but we're writing markdown here.